[FLINK-17342] enqueue savepoint requests in UC mode

rkhachatryan opened the pull request:
## What is the purpose of the change

The current behavior of forcing savepoints ignores max-concurrent-checkpoints limit. Which violates Unaligned Checkpoints (UC) assumption of a single concurrent barrier.

This PR changes `CheckpointCoordinator` behavior to enqueue savepoint requests instead of forcing them if the limit is reached (in UC mode). Supporting changes are to execute these requests after one of the pending checkpoints is completed and prioritize savepoints before checkpoints.

I consider this as a workaround because there is already work being done in `CheckpointCoordinator` and I'd like to minimize intersections.

The proper solution would be to introduce several separate queues and a single worker thread that processes requests from these queues: savepoint requests, (periodic) checkpoint requests, and ack/nack requests (thread-safe bounded O(1) queues). That would eliminate most of the complexity with callbacks and checks.

## Brief change log

1. Remove periodicTriggeringSuspended. Instead, if periodic request can't be executed now is ignored.
2. Check queue before executing a new request (not after).
3. Execute queued request when pending checkpoint is completed (instead of resuming timer).
4. Don't set CheckpointRequest.force in UC mode.
5. Change requests queue to PriorityQueue to prioritize savepoints. Limit its size.
6. Move all request-related decision logic into one place (and one sync block).
7. For savepoints, if running in UC mode and limit is reached then enqueue request instead of forcing it
8. Don't throw trigger exceptions. Currently, they are not reported to `CheckpointFailureManager` (and besides that, it would ignore them)



## Verifying this change

This change added tests and can be verified as follows:

 - Added `CheckpointCoordinatorTest.testSavepointScheduledInUnalignedMode` to test new CheckpointCoordinator behavior (savepoints in UC mode)
 - Added `CheckpointRequestDeciderTest` to test decision logic in isolation

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

pnowojski commented:
Travis failed with a deadlock in `CheckpointCoordinatorTriggeringTest`

rkhachatryan commented:
Hmm.. Travis was reporting some unrelated failures via API. Thanks for looking into it.
The problems were caused by mocks and the wrong `CheckpointCoordinator` configuration.
I've fixed them and also rebased onto the latest master, please take a look.

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointProperties.java:
```diff
@@ -248,6 +248,15 @@ public String toString() {
 			false);
 
+	private static final CheckpointProperties SYNC_SAVEPOINT_NOT_FORCED = new CheckpointProperties(
+			false,
```
Have you thought about implications of setting this flag to `false` in other places? like `PendingCheckpoint#canBeSubsumed()`? or `PendingCheckpoint#assertAbortSubsumedForced`?

rkhachatryan replied:
You're right, I missed this use of  the `force` flag.
Fixed by replacing `isForce` by `isSavepoint` in these methods.

pnowojski replied:
Have you added a test coverage for those bugs?

rkhachatryan replied:
This is covered by `PendingCheckpointTest.estCanBeSubsumed`.

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:
```diff
@@ -503,44 +507,17 @@ public boolean isShutdown() {
-		final CompletableFuture<CompletedCheckpoint> onCompletionPromise =
-			new CompletableFuture<>();
-		synchronized (lock) {
-			if (isTriggering || !triggerRequestQueue.isEmpty()) {
```
How was it working before? If savepoint was triggered while a checkpoint was enqueued/triggering, it should have triggered this branch and savepoint should enqueue after the checkpoint? Or am I missing something?

rkhachatryan replied:
Summarizing the offline discussion:
- if a request arrived while another is being *triggered* (or queue not empty) it is enqueued
- after the current request is executed (only triggering, not checkpoint completion), the queue is checked

Also, on completion of a checkpoint (not triggering) the queue is not checked; but the timer could be resumed which would result in checking the queue.

In this PR:
- the timer is not suspended at all
- instead, if `queue.head` can't be executed now it is either dropped (if periodic) or delayed (if savepoint)
- on checkpoint completion, the queue is checked

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointRequestDecider.java:
```diff
@@ -0,0 +1,156 @@
+	private final long minPauseBetweenCheckpoints;
+	private final Supplier<Integer> pendingCheckpointsSizeSupplier;
+	private final Function<Supplier<Optional<CheckpointTriggerRequest>>, Optional<CheckpointTriggerRequest>> sync;
+	private final PriorityQueue<CheckpointTriggerRequest> triggerRequestQueue = new PriorityQueue<>((r1, r2) -> {
+		if (r1.props.isSavepoint() != r2.props.isSavepoint()) {
+			return r1.props.isSavepoint() ? -1 : 1;
+		} else if (r1.props.forceCheckpoint() != r2.props.forceCheckpoint()) {
+			return r1.props.forceCheckpoint() ? -1 : 1;
+		} else if (r1.isPeriodic != r2.isPeriodic) {
+			return r1.isPeriodic ? 1 : -1;
+		} else {
+			return Long.compare(r1.timestamp, r2.timestamp);
+		}
+	});
```
I think providing a comparator is an idiomatic way of building a `PriorityQueue`. And then `r1` and `r2` names follow from it and the context of Request Decider (and also queue type).
The ordering is tested in `CheckpointRequestDeciderTest.testSavepointsPrioritized`.

rkhachatryan replied:
Extracted `checkpointTriggerRequestsComparator`.

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointRequestDecider.java:
```diff
@@ -0,0 +1,156 @@
+	private final Clock clock;
+	private final long minPauseBetweenCheckpoints;
+	private final Supplier<Integer> pendingCheckpointsSizeSupplier;
+	private final Function<Supplier<Optional<CheckpointTriggerRequest>>, Optional<CheckpointTriggerRequest>> sync;
```
I agree this type is quite complex. Reverted it to just `Object lock` as it currently doesn't provide benefits and synchronization will likely be removed soon.

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:
```diff
@@ -497,44 +504,17 @@ public boolean isShutdown() {
-		boolean advanceToEndOfTime,
-		CompletableFuture<CompletedCheckpoint> onCompletionPromise) {
-
+	private void startTriggeringCheckpoint(CheckpointTriggerRequest request) {
```
Extracted these changes among some others into a separate commit (initially, I tried to split it by logical changes but they all seem strongly connected).

rkhachatryan commented:
Thanks for the feedback @pnowojski .
I've updated the PR, please take a look.

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointRequestDecider.java:
```diff
@@ -0,0 +1,156 @@
+				return onCantCheckpointNow(candidate, TOO_MANY_CONCURRENT_CHECKPOINTS, () -> { /* not used */ }, isEnqueued);
+			} else if (nextTriggerDelayMillis(lastCompletionMs) > 0) {
+				return onCantCheckpointNow(candidate, MINIMUM_TIME_BETWEEN_CHECKPOINTS, () -> rescheduleTrigger.accept(nextTriggerDelayMillis(lastCompletionMs)), isEnqueued);
+			} else if (candidate != triggerRequestQueue.peek()) {
```
`else if (!isEnqueued)` to be consistent with other such checks? Or maybe introduce
```
boolean isAlreadyEnqueued(candidate) {
  return candidate == triggerRequestQueue.peek();
}
```
?

edit: or as I suggested below, drop the case of enqueued `candidate` altogether ?

rkhachatryan replied:
Changed code to always work through the queue.

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointRequestDecider.java:
```diff
@@ -0,0 +1,156 @@
+			Runnable postDrop,
+			boolean isEnqueued) {
+		if (candidate.props.forceCheckpoint()) {
+			return Optional.of(isEnqueued ? triggerRequestQueue.poll() : candidate);
```
it's a bit confusing the invartiant, that `candidate` can be still on the queue or not yet. I'm not sure, but It might cleaned up the code a bit, if you polled the candidate from the the queue immediately. This could unify some code paths with not enqueued candidate.

Especially that this `isEnqueued ? triggerRequestQueue.poll() : candidate` condition is already duplicated and can be easily forgotten in case of future changes.

rkhachatryan replied:
Yes, we could eliminate `isEnqueued` logic by either always adding (new) requests to the queue; or by always polling (enqueued) requests from it.
But it would have some performance penalty (`log(n)`). As it depends on the configuration I opted for a more complex code here.

rkhachatryan replied:
Changed code to always go through the queue.

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:
```diff
@@ -535,20 +540,16 @@ private void startTriggeringCheckpoint(CheckpointTriggerRequest request) {
-							timer);
+			final CompletableFuture<?> coordinatorCheckpointsComplete = pendingCheckpointCompletableFuture.thenComposeAsync(
+					pendingCheckpoint -> triggerAndAcknowledgeAllCoordinatorCheckpointsWithCompletion(coordinatorsToCheckpoint, pendingCheckpoint, timer),
+					timer);
```
Those lines should be wrapped somehow, also in other places after your changes lines are too long.

rkhachatryan replied:
As Flink guidelines don't name any specific limit, I usually assume that if it fits on my 14" screen then it could fit on most other devices:) 
I think it makes sense to split the line to attract attention to something important, but here it's just one lambda with one method call.

pnowojski replied:
Hmmm, indeed it doesn't say anything. At least not for java (for Scala it says 100).

Those lines are to wide for example for side-by-side comparisons (diffs) or when opening two files with vertical split.

rkhachatryan replied:
I think the code is more often read as it is, without comparison (and for comparison of long lines I use line-by-line format).
Please let me know if you prefer it to be wrapped anyways.

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:
```diff
@@ -43,7 +43,6 @@
 import org.apache.flink.runtime.state.StateBackend;
```
Generally speaking I'm not a big fan of this last commit, as it will make our lives more difficult to re-apply reverted refactoring commits. So I would actually prefer to drop/suspend it for now.

rkhachatryan replied:
As discussed offline, let me know if I should drop it (or feel free to drop it yourself).

rkhachatryan commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java:
```diff
@@ -969,15 +969,17 @@ private void configureCheckpointing() {
+				.setExactlyOnce(getCheckpointingMode(cfg) == CheckpointingMode.EXACTLY_ONCE)
+				.setPreferCheckpointForRecovery(cfg.isPreferCheckpointForRecovery())
+				.setTolerableCheckpointFailureNumber(cfg.getTolerableCheckpointFailureNumber())
+				.setUnalignedCheckpointsEnabled(cfg.isUnalignedCheckpointsEnabled())
```
I don't see any problems here, at least on `CheckpointCoordinator` side.
Do you have something specific in mind?

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointRequestDecider.java:
```diff
@@ -0,0 +1,156 @@
+			return r1.props.isSavepoint() ? -1 : 1;
+		} else if (r1.props.forceCheckpoint() != r2.props.forceCheckpoint()) {
+			return r1.props.forceCheckpoint() ? -1 : 1;
+		} else if (r1.isPeriodic != r2.isPeriodic) {
+			return r1.isPeriodic ? 1 : -1;
```
This is to prioritize requests submitted by the user (as opposed to scheduled periodically).

pnowojski replied:
Do we have a test coverage for that?

rkhachatryan replied:
Yes (added a test).

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointRequestDecider.java:
```diff
@@ -0,0 +1,156 @@
+	 */
+	private Optional<CheckpointTriggerRequest> chooseRequestToExecute(
+			boolean isEnqueued,
+			Supplier<CheckpointTriggerRequest> candidateSupplier,
```
This is to avoid extra `synchronized` when `peek`ing at queue head.

rkhachatryan replied:
Moved `synchronized` to the upper level (I prefer smaller (and less number of) `synchronized` sections, but it's not a strong preference).

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointRequestDecider.java:
```diff
@@ -0,0 +1,156 @@
+			long lastCompletionMs) {
+		return sync.apply(() -> {
+			final CheckpointTriggerRequest candidate = candidateSupplier.get();
+			if (candidate == null) {
+				return Optional.empty();
```
This check has to be done inside the `synchronized` section to avoid race conditions.

(I assume you meant `Optional.empty()`, not `@Nullable`; I don't see any `@Nullable` there).

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointRequestDecider.java:
```diff
@@ -0,0 +1,156 @@
+			if (triggerRequestQueue.size() < MAX_QUEUED_REQUESTS) {
+				triggerRequestQueue.offer(candidate);
+			} else {
+				candidate.onCompletionPromise.completeExceptionally(new CheckpointException(CheckpointFailureReason.TOO_MANY_CHECKPOINT_REQUESTS));
```
I introduced this check only to ease debugging in case of bugs (or probably misconfiguration). 
This shouldn't happen normally so I don't think testing it adds enough value.

pnowojski replied:
I would still add a test for that. It's a valid feature and handling the corner/error cases is an important thing.

rkhachatryan replied:
Added a test.

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointRequestDecider.java:
```diff
@@ -0,0 +1,156 @@
+
+@SuppressWarnings("ConstantConditions")
+class CheckpointRequestDecider {
+	private static final int MAX_QUEUED_REQUESTS = 1000;
```
Yes, it's a new limit.
Yes, previously, the queue would be cleaned, but only from periodic requests; user-submitted requests (i.e. savepoints) were executed in force mode.
Yes, requests can now be blocked for a longer time.

rkhachatryan commented:
Thanks for the feedback @pnowojski.

I've updated the code, please take a look.

Some changes not related directly to your comments:
1. I realized that with a queue limit in place we don't have to drop requests anymore if there are too many pending checkpoints. Dropping can cause a delay (until timer fires next time) so I removed it.
2. If queue is too big `decider` now drops the least significant queued request (instead of the new one). I replaced queue with a tree for that.
3.  Increased test coverage for these and existing cases.

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointRequestDecider.java:
```diff
@@ -188,4 +198,12 @@ int getNumQueuedRequests() {
+	private static void logInQueueTime(CheckpointTriggerRequest request) {
+		if (LOG.isDebugEnabled()) {
+			long timeInQueue = request.timestamp - currentTimeMillis();
+			if (timeInQueue > LOG_TIME_IN_QUEUE_THRESHOLD_MS) {
+				LOG.debug("checkpoint request time in queue: {}", timeInQueue);
+			}
```
bump to `100ms` and log as `INFO`? `DEBUG` wouldn't be visible in the most cases, (for example when users will be reporting to us that checkpoints are not being triggered).

rkhachatryan commented:
Fixed related issue by completely removing `timestamp` parameter (and the associated assertions in tests).

Unrelated issues:
https://issues.apache.org/jira/browse/FLINK-16921
https://issues.apache.org/jira/browse/FLINK-17336

pnowojski commented:
Thanks @rkhachatryan . I've fixed checkStyle failure and merged to master.
